text/0000-router-service.md

```diff
@@ -0,0 +1,440 @@
+// redirected to /person/0/detail. And /person/5 will get redirected to /person/0.
+```
+
+
+### Actively discourage use of private API
+
+This RFC provides public API for doing the things people have become accustomed to doing via private API. To eliminate confusion over the correct way, we should hide all the private API away behind symbols, and provide deprecation warnings per our usual release policy around breaking "widely-used private APIs".
+
+Some of the private APIs we should mark and warn include:
+
+ - transition.state
+ - transition.params
+ - `lookup('router:main')` (should use `service:router` instead)
+
+
+## Dynamically-Scoped Variables
+
+The next section introduces general-purpose, dynamically-scoped variables. Why am I doing this in "The Router Service RFC"? Because dynamically-scoped RouteInfo has several nice capabilities that will be introduced in the subsequent section. And once we have one dynamically-scoped variable, we have the option of exposing that capability for other uses.
+
+### In the general case
+
+A [dynamically-scoped variable](https://en.wikipedia.org/wiki/Scope_(computer_science)#Dynamic_scoping) takes its value from its calling context, as opposed to its lexical context. Since dynamically-scoped variables are powerful to the point of potential danger, their syntax is intended to be appropriately verbose.
+
+Reading a dynamically-scoped variable in handlebars:
+
+ ```hbs
+ {{!- retrieve the value of a dynamically scoped variable }}
+ (get-dynamic-variable "myVariableName")
+ ```
+
+Defining a component that reads a dynamically-scoped variable:
+
+```js
+let MyComponent = Ember.Component.extend({
+  didInsertElement() {
+    // Access to the variable here is intended to be indistinguishable
+    // from a normal, explicitly-passed input argument. 
+    doSomethingWith(this.get('myDynamicVariable'));
+  }
+});
+MyComponent.reopenClass({
+  // This is where the specialness happens.
+  getDynamicVariables: [ 'myDynamicVariable' ]
+});
+```
+
+Defining a helper that reads a dynamically-scoped variable:
+
+```js
+let MyHelper = Ember.Helper.extend({
+  compute(params, hash) {
+    return doSomethingWith(hash.myDynamicVariable);
+  }
+});
+MyHelper.reopenClass({
+  getDynamicVariables: [ 'myDynamicVariable' ]
+});
+```
+
+Setting a dynamically-scoped variable:
+
+```hbs
+{{#with-dynamic-variable "myVariableName" someValue}}
+  {{!-
+    within this block AND ALL ITS DESCENDANTS until
+    otherwise overridden by another with-dynamic-variable statement, 
+    `get-dynamic-variable "myVariableName"` returns someValue.
+  -}}
+{{/with-dynamic-variable}}
+```
+
+Note that there is no `set-dynamic-variable`. You can only introduce new scopes, not mutate your containing scope. There is also no way to set a dynamically-scoped variable directly from Javascript -- your component must use a `with-dynamic-variable` block within its handlebars template.
```

Comment:
These variables live in a completely separate namespace from normal (lexically-scoped) variables. `{{common}}` can never refer to a dynamically-scoped variable. You would need to write `{{get-dynamic-variable "common"}}`. 

So there is no shadowing issue between dynamic and lexical variables.

Dynamic variables in inner scopes can shadow dynamic variables in outer scopes, which is the point of `#with-dynamic-variable`